src/main/scala/is/hail/expr/Type.scala

```diff
@@ -643,6 +655,9 @@ final case class TAggregable(elementType: Type) extends TContainer {
 }
 
 abstract class TContainer extends Type {
+
+  var elementsRequired: Boolean = false
```

Comment:
This should be a def.

Reply:
We need a syntax for this in the parser.  I was thinking `Array[T*]` where `*` implies required.  Thoughts @danking @tpoterba?

Reply:
Sure, asterisk sounds OK to me

Reply:
Haskell uses `!` to indicate non-lazy things. I like that in our case because it's like "no really, a T, not a maybe T!". `Array[T!]`

Reply:
Yeah, I like !, that's less overloaded than * which makes me think of Seq. I was going to suggest + since that requires "at least one" in regex as opposed to *, but then it looks like covariant type.

Reply:
I'm starting not to like trailing ! because it will get lost when array element types are long, for example:

```
  Struct {
    ...,
    gs: Array[Struct {
      GT: Call,
      ...
      PL: Array[Int!]
  }!]
```

You don't see the bang until the end.  What about `Array[!T]`?  (Or should it be upside down bang, haha?)

I think the same goes for fields.  Should it be `Struct { foo: Int! }` or `Struct { foo: !Int }`.  I like the latter because it is a property of the field and not the type.

Reply:
Haskell also prefers the `!T` syntax. I have no preference.

I don't think of it as an intrinsic property of the field. I think of it as a property of the field by consequence of the field's type. If I load the `foo` field as an intermediate value, the intermediate value is also of type `!Int`.